mqtt-streaming/src/main/scala/akka/stream/alpakka/mqtt/streaming/impl/ClientState.scala

```diff
@@ -195,17 +201,28 @@ import scala.util.{Failure, Success}
 
   def disconnected(data: Disconnected)(implicit mat: Materializer): Behavior[Event] = Behaviors.receivePartial {
     case (context, ConnectReceivedLocally(connect, connectData, remote)) =>
+      import context.executionContext
+
       val (queue, source) = Source
-        .queue[ForwardConnectCommand](1, OverflowStrategy.dropHead)
-        .toMat(BroadcastHub.sink)(Keep.both)
-        .run()
-      remote.success(source)
+        .queue[ForwardConnectCommand](1, OverflowStrategy.backpressure)
+        .prefixAndTail(0)
+        .map(_._2)
+        .toMat(Sink.head)(Keep.both)
+        .run
```

Comment:
@ennru @2m We'd be interested in your views on this pattern as opposed to using `BroadcastHub`. We are wondering if it might be more efficient to use `prefixAndTail(0)` rather than a `BroadcastHub` where we know there is only one consumer - possibly less machinery. Note that this `prefixAndTail(0)` pattern is also used in the Unix Domain Socket code. WDYT?

Reply:
I'd suggest to use the more idiomatic `BroadcastHub` until that turns out to be an actual performance issue. But I can see how broadcast hub introduces a bit of extra baggage but it's probably not a problem.

Reply:
Everything is a problem in our resource constrained environment! ;-) we should of course benchmark it. 

Putting the question another way, do you see anything wrong with this approach in principle?

Reply:
I put together a silly comparison and found that at least in terms of running time `BroadcastHub` performs slightly better (~6% faster): https://github.com/longshorej/garbage/tree/master/akka-broadcast-comparison

```
akka-broadcast-comparison WARMUP: 2 * 100x
akka-broadcast-comparison ************************
akka-broadcast-comparison [broadcastHub] time (averaged over 100 times): 30.486356810000004ms
akka-broadcast-comparison [prefixAndTail] time (averaged over 100 times): 27.15035694ms
akka-broadcast-comparison [broadcastHub] time (averaged over 100 times): 25.54272718ms
akka-broadcast-comparison [prefixAndTail] time (averaged over 100 times): 26.1548963ms
akka-broadcast-comparison ************************
akka-broadcast-comparison RUN: 1000x
akka-broadcast-comparison ************************
akka-broadcast-comparison [broadcastHub] time (averaged over 1000 times): 25.498494859999997ms
akka-broadcast-comparison [prefixAndTail] time (averaged over 1000 times): 27.157771376ms
akka-broadcast-comparison ... finished with exit code 0
```

This is timing how long it takes to offer 10k items to the queue and consume them via a `Sink.ignore`. The result indicated is a mean average over 1k executions.

The first warmup having `prefixAndTail` slightly ahead is just due to everything JITing. By the second warmup, `BroadcastHub` performs a bit better.

Of course, YMMV given the variance in microbenchmarks. But it to me supports what @jrudolph is saying, i.e. let's use `BroadcastHub`.

Reply:
Here's a `/usr/bin/time` comparision for peak RSS, again slight advantage in most metrics to BroadcastHub. Also note again this is a pretty dumb benchmark :stuck_out_tongue: 

### BroadcastHub

```
/usr/bin/time -v java -jar /home/longshorej/work/garbage/akka-broadcast-comparison/target/scala-2.12/akka-broadcast-comparison-assembly-0.1.0-SNAPSHOT.jar broadcastHub
************************
WARMUP: 2 * 100x
************************
[broadcastHub] time (averaged over 100 times): 32.08024019ms
[broadcastHub] time (averaged over 100 times): 27.110958380000003ms
************************
RUN: 1000x
************************
[broadcastHub] time (averaged over 1000 times): 25.825699086ms
	Command being timed: "java -jar /home/longshorej/work/garbage/akka-broadcast-comparison/target/scala-2.12/akka-broadcast-comparison-assembly-0.1.0-SNAPSHOT.jar broadcastHub"
	User time (seconds): 111.35
	System time (seconds): 3.40
	Percent of CPU this job got: 354%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:32.33
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 235636
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 0
	Minor (reclaiming a frame) page faults: 34031
	Voluntary context switches: 857104
	Involuntary context switches: 170048
	Swaps: 0
	File system inputs: 0
	File system outputs: 0
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```

## prefixAndTail

```
/usr/bin/time -v java -jar /home/longshorej/work/garbage/akka-broadcast-comparison/target/scala-2.12/akka-broadcast-comparison-assembly-0.1.0-SNAPSHOT.jar prefixAndTail
************************
WARMUP: 2 * 100x
************************
[prefixAndTail] time (averaged over 100 times): 34.995869400000004ms
[prefixAndTail] time (averaged over 100 times): 27.3833828ms
************************
RUN: 1000x
************************
[prefixAndTail] time (averaged over 1000 times): 27.194013631ms
	Command being timed: "java -jar /home/longshorej/work/garbage/akka-broadcast-comparison/target/scala-2.12/akka-broadcast-comparison-assembly-0.1.0-SNAPSHOT.jar prefixAndTail"
	User time (seconds): 116.08
	System time (seconds): 3.65
	Percent of CPU this job got: 351%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:34.02
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 236244
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 0
	Minor (reclaiming a frame) page faults: 36818
	Voluntary context switches: 1029486
	Involuntary context switches: 193720
	Swaps: 0
	File system inputs: 0
	File system outputs: 0
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```

Reply:
BroadcastHub it is then! Thanks for the quick test.

Reply:
Have you considered `Sink.asPublisher`? Might be interesting to measure how that compares.

Reply:
Yes. And with `Sink.asPublisher[T](fanout: Boolean)` you can control wether you want to support fanout or not.

Reply:
Ah, that's quite a bit faster for the microbenchmark, which makes sense as it doesn't need to wait for an offer signal:

```
fanOut = false

[publisher] time (averaged over 1000 times): 3.76545015ms
```
```
fanOut = true
[publisher] time (averaged over 1000 times): 8.502225306ms
```

I'm not sure how we can apply it to the MQTT usage though, as we are offering items one at a time occasionally. Perhaps we can use `fanOut = true` and run a `Source.single` through it in each place where we currently offer. I think that means we lose any form of backpressure on the outgoing elements though.